### PR TITLE
Manually check protocol for http -> https redirect

### LINKED ISSUE #2363: HTTP redirects to HTTPS audit doesn't work on android
When you run Lighthouse on WPT or on any Android device, the security domain does not work properly and thus fails the HTTP redirect check, [see this report as example](https://www.webpagetest.org/lighthouse.php?test=170525_5S_14TQ&run=1). 

We should do what we did for is on HTTPS and just check for the scheme.

### PR DESCRIPTION
**Pull request by ev1stensberg**
fixes #2363

I’m not sure if I did this right. Seems too simple, but may be my lack of knowledge of what `afterPass` is doing behind the scenes.

### CODE REVIEW

**Comment by ev1stensberg** [comment]
Should I remove the old tests to test against old -> new urls accordingly?

File: lighthouse-core/gather/gatherers/http-redirect.js [hunk 1]
```diff
@@ -36,39 +36,9 @@ class HTTPRedirect extends Gatherer {
...
-    ]).then(result => {
-      // Clear timeout. No effect if it won, no need to wait if it lost.
-      clearTimeout(noSecurityChangesTimeout);
-      return result;
-    }).catch(err => {
-      clearTimeout(noSecurityChangesTimeout);
-      throw err;
-    });
+    return {
+      value: options.url.includes('https')
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/http-redirect.js]: maybe slightly more robust to bring in `lighthouse-core/libs/url-shim` and do `new URL(options.url).protocol === 'https:'`?

Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/http-redirect.js]: Actually tried to implement it as:

**Comment by ev1stensberg** [comment]
Proof of concept, adding tests in a sec! 
skjermbilde 2017-05-31 kl 00 31 56
skjermbilde 2017-05-31 kl 00 30 26

File: lighthouse-core/gather/gatherers/http-redirect.js [hunk 2]
```diff
@@ -27,48 +28,12 @@ class HTTPRedirect extends Gatherer {
 
   constructor() {
     super();
-    this._preRedirectURL = undefined;
-  }
-
-  beforePass(options) {
-    this._preRedirectURL = options.url;
-    options.url = this._preRedirectURL.replace(/^https/, 'http');
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/http-redirect.js]: we still need this part though, that's what lets us test the redirect when an https url is given :)

Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/http-redirect.js]: :((

**Comment by ev1stensberg** [comment]
Done now, I'd squash this 🥑  on acceptance!

**Comment by patrickhulce** [comment]
want to give `https://pwa.glitch.me/` a run too to make sure this fixes #2383 too? 🐦 🐦 💎

**Comment by ev1stensberg** [comment]
Can make a dive for it tomorrow, it's 1AM here 🌜

File: lighthouse-core/gather/gatherers/http-redirect.js [hunk 3]
```diff
@@ -36,39 +37,9 @@ class HTTPRedirect extends Gatherer {
...
-      timeoutPromise
-    ]).then(result => {
-      // Clear timeout. No effect if it won, no need to wait if it lost.
-      clearTimeout(noSecurityChangesTimeout);
-      return result;
-    }).catch(err => {
-      clearTimeout(noSecurityChangesTimeout);
-      throw err;
-    });
+    return {
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/http-redirect.js]: can remove all the `this._preRedirectURL` stuff up above, as well

Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/http-redirect.js]: Did, but @patrickhulce told me to let it stay, I'll look into it tomorrow 👍

Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/http-redirect.js]: @brendankenny how else will the URL to be tested be changed if they provide a URL already on `https`?

Comment (brendankenny) [inline lighthouse-core/gather/gatherers/http-redirect.js]: sorry, meant only `this._preRedirectURL` doesn't need to be a member variable anymore. Still need the URL change

**Comment by dandv** [comment]
Would this also fix http://m.uber.com being flagged as not redirecting to HTTPS, when in fact it is, via HSTS?

image

image

On the other hand, http://blueseed.com doesn't fail the redirect check, even though the same HSTS mechanism is used.

**Comment by ev1stensberg** [comment]
Only checks for `HTTPS` right now @dandv, should I integrate to check header for `HSTS` as well @patrickhulce @brendankenny ?

**Comment by ev1stensberg** [comment]
Works here. Tested on Google Pixel XL
skjermbilde 2017-05-31 kl 15 26 59

File: lighthouse-core/gather/gatherers/http-redirect.js (hunk 3 above)
Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/http-redirect.js]: Changed in new commit along with failing tests. I added some clauses to make dates work too

**Comment by ev1stensberg** [comment]
Fixes #2383 , but I need to fix the promise-race from the gross implementation in `driver`. Any suggestions? Also need to remove the security module from the mocks & src

**Comment by brendankenny** [comment]
Not sure what's going on in 3132c89. driver methods generally shouldn't navigate, that should be driven by gather-runner.

What was the implementation before the latest commit missing? JS redirects?

**Comment by ev1stensberg** [comment]
Yep. One thing we could do, is to add another `this.something` [here](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/gather/driver.js#L65) and assign the host url to that when the first network monitor starts, and later just fetch it with a util that has the same implementation as the one I just committed. This way, there's no promise race or anything, just the same logic we've used in the previous commits. 

Also, if we decide to make a method like the one I just implemented, it's easier to look for headers and other minor bugs that might occur in hindsight. 

I dislike the approach I made in this commit, as the promise-race makes the runtime 👎 . Good side is however, that we could do what I just mentioned above. 

Also experimented a bit with checking for SW'ers, we could do that, but that is a one-time bug that could be dropped if we go with what I initially described.

**Comment by ev1stensberg** [comment]
`this.something` could be reassigned [here](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/gather/driver.js#L535) to match the initial route

**Comment by paulirish** [comment]
fwiw here is the diff before the latest commit: https://github.com/GoogleChrome/lighthouse/compare/3132c8915230f1131ccfc59279df8ba4f7ca6fe8~

**Comment by brendankenny** [comment]
hmm, @ev1stensberg, maybe I don't understand what you're trying to do. From `getRedirectURL`, it appears that it's trying to navigate to the http page, and watches the completed network requests looking for the https url. Is that correct?

If that's the case, it's probably easier to look at the `networkRecords` for that pass, which should already have all this data. That's available in `afterPass` (see e.g. [`response-compression`](https://github.com/GoogleChrome/lighthouse/blob/400457b22483510badbd3394db7fc6e335237997/lighthouse-core/gather/gatherers/dobetterweb/response-compression.js#L62) as an example of use there).

But I'm not sure if that'll fix all our issues:
1) I don't actually know if redirecting in js should pass the check. It's arguably as secure as an http redirect (or even HSTS if the first visit to the site is over http), I guess, so maybe it's fine.
2) Should the https redirect have to take the user back to the same exact url with the more secure protocol? That seems stricter than necessary.
3) we run the `http-redirect` gatherer [in the same pass as `html-without-javascript`](https://github.com/GoogleChrome/lighthouse/blob/400457b22483510badbd3394db7fc6e335237997/lighthouse-core/config/default.js#L50-L51), so I don't believe the js to do the redirect will run anyways.
4) Why aren't HSTS redirects being picked up by our [redirect monitor](https://github.com/GoogleChrome/lighthouse/blob/400457b22483510badbd3394db7fc6e335237997/lighthouse-core/gather/driver.js#L536-L542)?

These seem complicated enough that they're worth figuring out on their own. In the meantime, it would be great to get the fix in for #2363 before figuring all that out.

**Comment by ev1stensberg** [comment]
Should I revert the latest commit & pick up on a separate PR?

**Comment by brendankenny** [comment]
> Should I revert the latest commit & pick up on a separate PR?

SG, but looks like you already did :)

**Comment by ev1stensberg** [comment]
Yea, common sense 😺

**Comment by ev1stensberg** [comment]
This is ready for the first roud now I assume, will implement what we've discussed / the idea I had tomorrow or so!

**Comment by ev1stensberg** [comment]
This is ready now, we had to update the URLs after waiting approx 5secs for full load. Implemented `this._url` so that tests also can pick up the context.

**Comment by ev1stensberg** [comment]
So much stress for a simple change :(

**Comment by ev1stensberg** [comment]
Hm, weird , Unit Test still failing, worked locally. Lemme get this finalized tomorrow. This module sure is fine-grained 🎭

**Comment by ev1stensberg** [comment]
So I can repro the Unit test, as the `setTimeout` isn't valid in Mocha 👎  Could we leave the `afterPass` tests for the smoke tests? Would be able to remove the `this._url` too and use the prev implementation of `options.url`
skjermbilde 2017-06-01 kl 12 23 44

**Comment by ev1stensberg** [comment]
So I tested http://jakearchibald.github.io/svgomg/ which should redirect after load. However, `httpsRedirect` doesn't pick this up. I'll reimplement this entire PR, as it's not working as expected, way to fragile. I'll be using `Network` to view page redirects to check against the URL we were given in `httpsRedirect`

**Comment by ev1stensberg** [comment]
Should fix https redirect issues

**Comment by ev1stensberg** [comment]
Signing this of @patrickhulce @brendankenny @paulirish 👍  Fixes @dandv 's issue, #2383 & #2363

**Comment by ev1stensberg** [comment]
skjermbilde 2017-06-01 kl 22 02 55
Anything I can do to make sure it doesn't happen, or should we wait for #2402 ?

**Comment by brendankenny** [comment]
> Anything I can do to make sure it doesn't happen, or should we wait for #2402 ?

yes, that will be fixed by #2402

**Comment by ev1stensberg** [comment]
Gotcha, sorry for the super-embarrasing jrdev mistakes! This is ready on my end, I removed tests for `afterPass`, as the tests didn't pick up `options.driver._httpsArr`, should be bullet-proof through smoketests though!

File: lighthouse-core/gather/gatherers/http-redirect.js [hunk 4]
```diff
@@ -36,39 +36,20 @@ class HTTPRedirect extends Gatherer {
...
-    const timeout = options._testTimeout || 10000;
-
-    const securityPromise = options.driver.getSecurityState()
-      .then(state => {
-        return {
-          value: state.schemeIsCryptographic
-        };
+    let isRedirect = false;
+
+    if(options.driver._httpsArr.length !== 0) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/http-redirect.js]: I'm not really understanding why this is necessary @ev1stensberg could you elaborate? As @brendankenny mentioned we're already saving every network request that is issued (which will have flags for redirect, status code, scheme, etc) that you could check in `options`. I'm also interested why jakearchibald's site did not work with the previous approach, the `afterPass` is already quite some time after load and the page has settled, maybe we should just fix the redirect monitor instead if we're not picking up those changes :)

UPDATE: oh wait, does he do a `location = ...` too? we probably only examine redirects right now

File: lighthouse-core/gather/gatherers/http-redirect.js [hunk 5]
```diff
@@ -36,39 +36,20 @@ class HTTPRedirect extends Gatherer {
...
+      const httpsExpectURL = options.url.replace(/^http/, 'https');
+      options.driver._httpsArr.forEach( (httpsUrl) => {
+        if(httpsUrl.url === httpsExpectURL) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/http-redirect.js]: I'd say we don't want to enforce that they redirect to the exact entered URL just with a scheme change, for example when testing `http://theverge.com` it ultimately sends you to `https://www.theverge.com/` which seems fine. @brendankenny did you have a different scenario in mind that you wanted exact equality?

File: lighthouse-core/gather/gatherers/http-redirect.js (hunk 4 above)
Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/http-redirect.js]: Smoketests are failing if it's not checked for, as it results to `undefined`. And yeah, he also is I think. Is there something I need to change here?

File: lighthouse-core/gather/gatherers/http-redirect.js (hunk 5 above)
Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/http-redirect.js]: I formed this after `http://m.uber.com` fails to resolve to https on the prev approach. With `'/'`, we are assuming the root route is https if it's in the arr. Anything I need to change here?

**Comment by ev1stensberg** [comment]
Left comments, for the removed tests, I can try to wrap a class to make them pass instead of removing. Is there anything I need to change that I've failed to implement? / need change?

**Comment by ev1stensberg** [comment]
`options.url` didn't pick up the latest URL through `location` change. If we default to looking at network records, we're now checking if we are on a http protocol for `options.url`, if that is the case, we have to look for a `https` url that matches the one that we've got in `httpsArr`. If we're at https already, this should be picked up by network records anyway, so we just have to test the two urls against each other. It might be that we cannot use protocol to check, as pages could be `m.uber.com`, so we're checking if we're on a root website if this happens. 

Also, in the latest commit, I've added a protocol check before looking at network, so that we can fetch if `options.url` has already been updated. This way, we're covering a lot of cases.

**Comment by paulirish** [comment]
@ev1stensberg are you pushing only to trigger tests?

you can do any of these on your own fwiw:

```sh
yarn unit-core
yarn test
yarn smoke
yarn smoketest
```

**Comment by ev1stensberg** [comment]
The `noopsi` commit was to try to see if the tests were buggy at the CI, I usually run those before I commit. Sorry for the excessive commits.. #jrdev

**Comment by paulirish** [comment]
Closing this as I'm going to take a stab at it based on the plan in https://github.com/GoogleChrome/lighthouse/issues/2363#issuecomment-313768112

Even, Thanks again for trying to make this work!

**Comment by ev1stensberg** [comment]
No worries, good luck!

**Pull request closed without merging**